gadgets/trace_ssl/program.bpf.c

```diff
@@ -0,0 +1,432 @@
+//
+// Based on sslsniff from BCC by Adrian Lopez
+// https://github.com/iovisor/bcc/blob/master/tools/sslsniff.py
+//
+// Based on sslsnoop from bpftrace by Tao Xu
+// https://github.com/bpftrace/bpftrace/blob/master/tools/sslsnoop.bt
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+#include <gadget/buffer.h>
+#include <gadget/macros.h>
+#include <gadget/mntns_filter.h>
+
+#define MAX_BUF_SIZE 8192
+#define MAX_ENTRIES 8192
+
+enum operation {
+	/* libssl */
+	libssl_SSL_do_handshake,
+	libssl_SSL_read,
+	libssl_SSL_write,
+
+	/* libgnutls */
+	libgnutls_gnutls_record_send,
+	libgnutls_gnutls_record_recv,
+
+	/* libnss */
+	libnss_PR_Write,
+	libnss_PR_Send,
+	libnss_PR_Read,
+	libnss_PR_Recv,
+
+	/* libcrypto */
+	libcrypto_rsa_ossl_public_encrypt,
+	libcrypto_rsa_ossl_public_decrypt,
+	libcrypto_rsa_ossl_private_encrypt,
+	libcrypto_rsa_ossl_private_decrypt,
+	libcrypto_RSA_sign,
+	libcrypto_RSA_verify,
+	libcrypto_ossl_ecdsa_sign,
+	libcrypto_ossl_ecdsa_verify,
+	libcrypto_ossl_ecdh_compute_key,
+};
+
+struct event {
+	gadget_mntns_id mntns_id;
+	enum operation operation;
+	gadget_timestamp timestamp;
+	u64 latency_ns;
+	u32 pid;
+	u32 tid;
+	u32 uid;
+	u32 gid;
+	u32 len;
+	u64 retval;
+	char comm[TASK_COMM_LEN];
+	u8 buf[MAX_BUF_SIZE];
+};
+
+#define BASE_EVENT_SIZE ((size_t)(&((struct event *)0)->buf))
+#define EVENT_SIZE(X) (BASE_EVENT_SIZE + ((size_t)(X)))
+
+GADGET_TRACER_MAP(events, 1024 * 256);
+GADGET_TRACER(ssl, events, event);
+
+const volatile bool record_data = true;
+GADGET_PARAM(record_data);
+
+/* used for context between uprobes and uretprobes of all operations */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, u64); // mntns_id
+} mntns_context SEC(".maps");
+
+/* used for context between uprobes and uretprobes of ssl operations */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, u64); // timestamp
+} ssl_starttime SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, void *); // buffer
+} ssl_bufs SEC(".maps");
+
+/* used for context between uprobes and uretprobes of libcrypto operations */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, u64); // timestamp
+} crypto_starttime SEC(".maps");
+
+/**
+ * clean up the maps when a thread terminates,
+ * because there may be residual data in the map
+ * if a userspace thread is killed between a uprobe and a uretprobe
+ */
+SEC("tracepoint/sched/sched_process_exit")
+int trace_sched_process_exit(void *ctx)
+{
+	u32 tid;
+
+	tid = (u32)bpf_get_current_pid_tgid();
+	bpf_map_delete_elem(&mntns_context, &tid);
+	bpf_map_delete_elem(&ssl_starttime, &tid);
+	bpf_map_delete_elem(&ssl_bufs, &tid);
+	bpf_map_delete_elem(&crypto_starttime, &tid);
```

Comment:
Optional suggestion: we now have several maps indexed by the tid. I think it would be cleaner to have a single map with a struct type as value.